Create bounties for the runtime.

shamil-gadelshin opened the pull request:
#### Comments
- new `pallet_bounty` was created and added to the runtime
- bounty creation, cancellation, veto, funding are supported
- bounty creation is possible for both a member and a council
- work submission, judgment, funds withdrawal are not supported yet
- tests and benchmarks are ready for an existing code
- council budget manager (helper) was created

#### Extrinsics
- create_bounty
- cancel_bounty
- veto_bounty
- fund_bounty


[Main issue](https://github.com/Joystream/joystream/issues/1998)

shamil-gadelshin committed:
runtime: Create bounty pallet.

shamil-gadelshin committed:
runtime: bounty: Add tests.

shamil-gadelshin committed:
runtime: bounty: Add oracle and contract types.

shamil-gadelshin committed:
runtime: bounty: Add the module comment.

shamil-gadelshin committed:
runtime: bounty: Add benchmarks.

shamil-gadelshin committed:
runtime: bounty: Add weights.

shamil-gadelshin committed:
runtime: bounty: Add min-max funding amount constraint.

shamil-gadelshin committed:
runtime: bounty: Add ‘cancel_bounty’ extrinsic.

shamil-gadelshin committed:
runtime: Add balance slashing for a cherry.

shamil-gadelshin committed:
runtime: bounty: Add veto_bounty extrinsic.

shamil-gadelshin committed:
runtime: bounty: Update benchmarks.

shamil-gadelshin committed:
runtime: bounty: Add creator_funding field to the Bounty.

shamil-gadelshin committed:
runtime: bounty: Refactor bounty creation.

shamil-gadelshin committed:
runtime: bounty: Add fund_bounty extrinsic.

shamil-gadelshin committed:
runtime: bounty: Add fund_bounty checks.

shamil-gadelshin committed:
runtime: bounty: Add MaxFundingReached event.

shamil-gadelshin committed:
runtime: bounty: Save fundings in the storage.

conectado commented on runtime-modules/bounty/src/benchmarking.rs:
```diff
@@ -0,0 +1,274 @@
+    verify {
+        let bounty_id: T::BountyId = 1u32.into();
+
+        assert!(Bounties::<T>::contains_key(bounty_id));
```
We could add an assertion that `BountyCount` also mutated as expected.

The same for `create_bounty_by_,member`

shamil-gadelshin replied:
Ok

shamil-gadelshin commented:
@bedeho 

The Substrate `treasury` pallet has [an implementation](https://github.com/paritytech/substrate/blob/c7f52d184727b11171eea458c4f277bc6b37df02/frame/treasury/src/lib.rs#L351) of your idea of bounty accounts.
I consider using it in the next PR of pallet_bounty.

conectado reviewed (commented):
Only left one comment and also this:

I think we should add an assertion after each bounty is created that the bounty exists as we expect in the storage just because it's easier to read from the assertions the pre-conditions for benchmarking.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+    Open,
+
+    /// Only specific members can submit the work.
+    Closed(Vec<MemberId>),
```
Closed was originally intended to mean that only bounty creator can submit, and thus also get any cherry back if funding fails. In this version where any set of third parties can make entry, I assume creator still gets cherry back if funding fails?

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+    pub contract_type: AssuranceContractType<MemberId>,
+
+    /// Bounty creator member ID, should be None if created by a council.
+    pub creator_member_id: Option<MemberId>,
```
Can we make this slightly more explicit so something like using a type

```Rust
enum BountyCreator {
Council,
Member(MemberId)
}
```

shamil-gadelshin replied:
Done.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+/// Crowdfunded bounty record.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct BountyRecord<Balance, BlockNumber, MemberId> {
```
Is there a reason we cannot just call this `Bounty` directly?

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct BountyRecord<Balance, BlockNumber, MemberId> {
+    /// Bounty creation parameters.
+    pub creation_params: BountyParameters<Balance, BlockNumber, MemberId>,
```
Are you 100% sure that every single value in `BountyParameters` is indeed required to be stored?

As the bounty progresses through the different stages an increasing number of the values become redundant, but its not that big of a problem. Addressing that would require lifting out smaller and smaller subsets into the different stage variants.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+    pub stage: BountyStage<BlockNumber>,
+
+    /// The current unspent "cherry" balance.
+    pub cherry_pot: Balance,
```
What does this do and why does it live here? What is meant by "unspent".

shamil-gadelshin replied:
Removed.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+    pub cherry_pot: Balance,
+
+    /// The current unspent funding balance.
+    pub current_funding: Balance,
```
What does this do and why does it live here? What is meant by "unspent".

shamil-gadelshin replied:
Renamed.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+    }
+}
+
+decl_event! {
```
Can we make all of these have exhaustive parameters, so full extrinsic argument list.

shamil-gadelshin replied:
Done.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+    pub creation_params: BountyParameters<Balance, BlockNumber, MemberId>,
+
+    /// The current bounty stage.
+    pub stage: BountyStage<BlockNumber>,
```
I don't think its possible to represent the stage in state like this in combination with

1. No iteration over bounties
2. Block height deadlines.

I have not looked through details, but something must fail here I think. I think the mildest possible "failure" would be that the stored `stage` value is not in synch with reality until some extrinsic occurs where its resynched. The right approach here should be that the stage itself is not stored, but just computable from a combination of when bounty was created, period deadlines and later milestone block heights.

shamil-gadelshin replied:
Refactored.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum BountyStage<BlockNumber> {
+    /// Bounty founding stage with starting block number.
+    Funding(BlockNumber),
```
What do you mean by "starting block", are you saying a bounty can start accepting funding in a future block, but not right away? That was not part of the spec I think. Can it still be cancelled here?

Also, are you not losing this information now in later stages? that would seem to make it impossible to compute stage without advancing it through `on_finalized` computation.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,619 @@
+    Funding(BlockNumber),
+
+    /// A bounty was canceled.
+    Canceled,
```
I do not think  cancel, veto are really distinct stages as far as the runtime is concerned, the only thing that matters when these actions are initiated is whether to go to a perpetual withdraw stage where funders can get their money back, or just to drop the bounty from the system right away if there are no such contributions that need to be recuperated.

shamil-gadelshin replied:
Removed.

bedeho reviewed (changes requested):
Promising start, but I think you need to think more deeply about how to organise the state of bounties to support our goals of no iteration and also no archival state. Here it seems you have come up with a state model with only very first stage in mind, and I do not think it works.

shamil-gadelshin committed:
runtime: bounty: Update benchmarks.

shamil-gadelshin committed:
runtime: bounty: Fix review comments.

1. Separated bounty state from the computed bounty stage.
2. Removed veto bounty state.
3. Added more info to the events.
4. Removed the cherry_pot field.

shamil-gadelshin committed:
runtime: bounty: Rename BountyStateInfo.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,676 @@
+
+    /// Bounty current state. It represents fact known about the bounty, eg.:
+    /// it was canceled or max funding amount was reached.
+    pub state: BountyMilestone<BlockNumber>,
```
Rename field to `milestone`.

shamil-gadelshin replied:
Done

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,676 @@
+                creator.clone(),
+            )?;
+
+            ensure!(
```
1. we will need to safely recover bounties in multiple places, so lets refactor this check+get into `ensure_bounty_exists`.
2. Let us avoid this mutable, this seems avoidable at little cost.

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,676 @@
+        pub Bounties get(fn bounties) : map hasher(blake2_128_concat) T::BountyId => Bounty<T>;
+
+        /// Double map for bounty funding. It stores member funding for bounties.
+        pub Funding get(fn funding_by_bounty_by_member): double_map hasher(blake2_128_concat)
```
Can we rename `Funding` to `BountyContributions`

shamil-gadelshin replied:
Done

bedeho commented on runtime-modules/bounty/src/lib.rs:
```diff
@@ -0,0 +1,676 @@
+            bounty_id: T::BountyId,
+            amount: BalanceOf<T>
+        ) {
+            let account_id = T::MemberOriginValidator::ensure_member_controller_account_origin(
```
We only allow contributing from controller account? I am fine with that, just confirming that this is what we wanted.

shamil-gadelshin replied:
Yes. Renamed to `controller_account_id` for clarity.

bedeho reviewed (changes requested):
Excellent work, you nailed fixing the state model here.

Not sure what we do about the bounty contribution cleanup in the success case, which presumably is the majority of cases. We may just have to accept it.

One alternative is to take k*`withdraw_fee` away from the initial contribution, and return it in exchange for cleanup, for some integer k>1, as an incentive to cleanup.

conectado reviewed (approved):
LGTM

shamil-gadelshin committed:
runtime: bounty: Refactoring.

bedeho reviewed (approved):
Excellent.
